documentation/src/algebraic-datatypes.md

```diff
@@ -0,0 +1,115 @@
+# Algebraic Datatypes
+
+Algebraic datatypes are also called discriminated unions, tagged unions, and sometimes "or types" colloquially. 
+They can be found in many functional languages (even if they're not named algebraic datatypes specifically), and have become widely known and used in Typescript 2.0.
+
+## Algebraic datatypes in C#
+
+While we don't have language support for algebraic datatypes in C#, there are some patterns on how you can implement something that behaves like one.
+
+We recommend the usage of the following pattern using an abstract base class with a private constructor and inner, deriving classes and usually a match function.
+This has the advantage that the algebraic options are namespaced as the inner class, which simplifies the naming.
+
+When your intention is to write something like an enum, but with attachable data to every option, algebraic datatypes is what you're looking for.
+
+## Example 1 (`ServiceStatus`)
+
+Our example has 3 states - a service can be stopped, starting or running. Each state also provides the `StandardOutput` and the `StandardError`.
+Our match method allows us to define what should happen for each configured option.
+The upside of that over a switch expression is that we have no problems with exhaustability (see Example 2).
+
+```csharp
+public abstract class ServiceStatus
+{
+	private ServiceStatus(Process? process)
+	{
+		StandardOutput = process?.StandardOutput.ReadToEnd() ?? string.Empty;
+		StandardError = process?.StandardError.ReadToEnd() ?? string.Empty;
+	}
+
+	public string StandardOutput { get; }
+
+	public string StandardError { get; }
+
+	public abstract void Match(

```

Comment:
Document that if the type is public (in a library) and the author wants to have the liberty of adding a new variant without breaking backwards compatibility, they should mark this method as `internal`.

Reply:
This is similar to what `#[non_exhaustive]` does in rust

Reply:
Added section in cf76766